src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs

```diff
@@ -50,12 +50,13 @@ IEnumerable<string> ExpandPageHierarchy()
                     // but only up to the pages root.
                     //
                     // This is easy because the 'page' token already trims the root directory.
-                    var end = context.PageName.Length;
+                    var path = pageActionDescriptor.ViewEnginePath;
```

Comment:
This turned out to be interesting. The demo we did happened to work because the Shared directory was the shared directory for the Area i.e. `/Areas/{2}/Pages/Shared/{0}.cshtml`. 

In the case of a nested dir such as `/Areas/{2}/Pages/Dir1/Shared/{0}.cshtml`, the `PageName` is `/Dir1/MyPage`, so we don't look at `/Dir1/MyPage/Shared`. Making `ViewEnginePath` and `PageName` distinct in the Shared case, so that `ViewEnginePath = "/Dir1/MyPage/Shared/MyPage"` resolves this concern.